File: src/features/accounts/repositories/accounts.repository.ts
import {
  supabase,
  type TypedSupabaseClient,
} from "../../../lib/supabase/client"
import {
  requireAuthenticatedUserId,
  requireQueryData,
} from "../../../lib/supabase/repository"
import { RepositoryError } from "../../../lib/supabase/types"
import type {
  AccountSummary,
  Decimal,
  TableUpdate,
} from "../../../lib/supabase/types"
import { normalizeDecimal } from "@/lib/financial-calculations/decimal"

export type CreateAccountInput = {
  accountTypeCode: string
  name: string
  currencyCode: string
  openingBalance?: Decimal
  notes?: string | null
  bankSubtype?: "debit" | "credit" | null
  creditCardLimit?: Decimal | null
  dueDayOfMonth?: number | null
  investmentType?: "stock_etf" | "crypto" | "other" | null
  balanceGrams?: Decimal | null
  propertyType?: "apartment" | "villa" | "land" | "office" | "other" | null
  ownershipPercentage?: Decimal | null
  businessType?: string | null
  industry?: string | null
  location?: string | null
  valuationAmount?: Decimal
  valuedOn?: string
  valuationMethod?: string | null
  valuationNotes?: string | null
  metalType?: "gold" | "silver" | null
  purity?: string | null
  purchaseDate?: string | null
  costPerUnit?: Decimal | null
}

export type UpdateAccountInput = {
  accountTypeCode?: string
  name?: string
  currencyCode?: string
  openingBalance?: Decimal
  notes?: string | null
  isActive?: boolean
  bankSubtype?: "debit" | "credit" | null
  creditCardLimit?: Decimal | null
  dueDayOfMonth?: number | null
  investmentType?: "stock_etf" | "crypto" | "other" | null
  balanceGrams?: Decimal | null
  propertyType?: "apartment" | "villa" | "land" | "office" | "other" | null
  ownershipPercentage?: Decimal | null
  businessType?: string | null
  industry?: string | null
  location?: string | null
  metalType?: "gold" | "silver" | null
  purity?: string | null
  purchaseDate?: string | null
  costPerUnit?: Decimal | null
}

export type AccountLifecycleEligibility = {
  accountId: string
  canClose: boolean
  closeBlockReason: string | null
  canDelete: boolean
  deleteBlockReason: string | null
  hasFinancialHistory: boolean
}

const immutableCurrencyMessage =
  "This account already contains financial history. Its currency cannot be changed."
const immutableOpeningBalanceMessage =
  "This account already contains financial history. Its opening balance cannot be changed."
const duplicateMetalAccountMessage =
  "You already have this type of Gold/Silver account in this currency. Go to that account and add a purchase instead of creating a new one."
const duplicateAccountNameMessage =
  "You already have an account with this name. Choose a different name."

type DatabaseError = {
  code?: string
  message: string
}

const accountSelect =
  "id,user_id,account_type_code,name,currency_code,opening_balance::text,notes,is_active,bank_subtype,credit_card_limit::text,due_day_of_month,investment_type,balance_grams::text,property_type,ownership_percentage::text,initial_ownership_percentage::text,closed_on,closed_reason,business_type,industry,location,metal_type,purity,purchase_date,cost_per_unit::text,created_at,updated_at" as const

export function requireAccountDecimalText(
  value: unknown,
  field: string,
  operation: string
): Decimal {
  if (typeof value !== "string" || normalizeDecimal(value) === null) {
    throw new RepositoryError({
      code: "database_error",
      message: `Account field ${field} must be a PostgreSQL decimal string`,
      operation,
    })
  }
  return value
}

function requireNullableAccountDecimalText(
  value: unknown,
  field: string,
  operation: string
): Decimal | null {
  return value === null
    ? null
    : requireAccountDecimalText(value, field, operation)
}

function mapAccountSummary(
  row: AccountSummary,
  operation: string
): AccountSummary {
  return {
    ...row,
    opening_balance: requireAccountDecimalText(
      row.opening_balance,
      "opening_balance",
      operation
    ),
    credit_card_limit: requireNullableAccountDecimalText(
      row.credit_card_limit,
      "credit_card_limit",
      operation
    ),
    balance_grams: requireNullableAccountDecimalText(
      row.balance_grams,
      "balance_grams",
      operation
    ),
    ownership_percentage: requireNullableAccountDecimalText(
      row.ownership_percentage,
      "ownership_percentage",
      operation
    ),
    initial_ownership_percentage: requireNullableAccountDecimalText(
      row.initial_ownership_percentage ?? null,
      "initial_ownership_percentage",
      operation
    ),
    cost_per_unit: requireNullableAccountDecimalText(
      row.cost_per_unit,
      "cost_per_unit",
      operation
    ),
  }
}

function throwAccountConstraintError(
  error: DatabaseError | null,
  operation: string
): void {
  if (!error) {
    return
  }

  if (
    error.code === "23505" &&
    error.message.includes("financial_accounts_user_currency_metal_type_key")
  ) {
    throw new RepositoryError({
      code: "conflict",
      message: duplicateMetalAccountMessage,
      operation,
      cause: error,
    })
  }

  if (
    error.code === "23505" &&
    error.message.includes("financial_accounts_non_metal_user_name_lower_key")
  ) {
    throw new RepositoryError({
      code: "conflict",
      message: duplicateAccountNameMessage,
      operation,
      cause: error,
    })
  }

  if (
    error.code === "23514" &&
    error.message.includes(immutableCurrencyMessage)
  ) {
    throw new RepositoryError({
      code: "constraint_violation",
      message: immutableCurrencyMessage,
      operation,
      cause: error,
    })
  }

  if (
    error.code === "23514" &&
    error.message.includes(immutableOpeningBalanceMessage)
  ) {
    throw new RepositoryError({
      code: "constraint_violation",
      message: immutableOpeningBalanceMessage,
      operation,
      cause: error,
    })
  }
}

export class AccountsRepository {
  private readonly client: TypedSupabaseClient

  constructor(client: TypedSupabaseClient = supabase) {
    this.client = client
  }

  async getAccounts(): Promise<AccountSummary[]> {
    const operation = "accounts.getAccounts"
    const userId = await requireAuthenticatedUserId(this.client, operation)
    const { data, error } = await this.client
      .from("financial_accounts")
      .select(accountSelect)
      .eq("user_id", userId)
      .order("created_at", { ascending: false })

    return requireQueryData(data, error, operation).map((row) =>
      mapAccountSummary(row, operation)
    )
  }

  async getAccount(id: string): Promise<AccountSummary> {
    const operation = "accounts.getAccount"
    const userId = await requireAuthenticatedUserId(this.client, operation)
    const { data, error } = await this.client
      .from("financial_accounts")
      .select(accountSelect)
      .eq("id", id)
      .eq("user_id", userId)
      .single()

    return mapAccountSummary(
      requireQueryData(data, error, operation),
      operation
    )
  }

  async createAccount(input: CreateAccountInput): Promise<AccountSummary> {
    const operation = "accounts.createAccount"
    if (input.accountTypeCode === "real_estate" || input.accountTypeCode === "business") {
      const { data, error } = await this.client.rpc("create_valued_account", {
        p_account_type_code: input.accountTypeCode,
        p_name: input.name,
        p_currency_code: input.currencyCode,
        p_property_type: input.propertyType ?? null,
        p_business_type: input.businessType ?? null,
        p_industry: input.industry ?? null,
        p_ownership_percentage: input.ownershipPercentage ?? "100",
        p_location: input.location ?? null,
        p_account_notes: input.notes ?? null,
        p_valuation_amount: input.valuationAmount ?? "0",
        p_valued_on: input.valuedOn ?? new Date().toISOString().slice(0, 10),
        p_valuation_method: input.valuationMethod ?? null,
        p_valuation_notes: input.valuationNotes ?? null,
      })
      throwAccountConstraintError(error, operation)
      const createdAccount = requireQueryData(data, error, operation)
      // PostgreSQL numerics in an RPC composite return may be JSON numbers. Re-read
      // through accountSelect so all decimal fields retain the repository's string contract.
      return this.getAccount(createdAccount.id)
    }
    const userId = await requireAuthenticatedUserId(this.client, operation)
    const { data, error } = await this.client
      .from("financial_accounts")
      .insert({
        user_id: userId,
        account_type_code: input.accountTypeCode,
        name: input.name,
        currency_code: input.currencyCode,
        opening_balance: input.openingBalance,
        notes: input.notes,
        bank_subtype: input.bankSubtype,
        credit_card_limit: input.creditCardLimit,
        due_day_of_month: input.dueDayOfMonth,
        investment_type: input.investmentType,
        balance_grams: input.balanceGrams,
        property_type: input.propertyType,
        ownership_percentage: input.ownershipPercentage,
        business_type: input.businessType,
        industry: input.industry,
        location: input.location,
        metal_type: input.metalType,
        purity: input.purity,
        purchase_date: input.purchaseDate,
        cost_per_unit: input.costPerUnit,
      })
      .select(accountSelect)
      .single()

    throwAccountConstraintError(error, operation)
    return mapAccountSummary(
      requireQueryData(data, error, operation),
      operation
    )
  }

  async updateAccount(
    id: string,
    input: UpdateAccountInput
  ): Promise<AccountSummary> {
    const operation = "accounts.updateAccount"
    const userId = await requireAuthenticatedUserId(this.client, operation)
    const update: TableUpdate<"financial_accounts"> = {}

    if (input.accountTypeCode !== undefined) {
      update.account_type_code = input.accountTypeCode
    }
    if (input.name !== undefined) {
      update.name = input.name
    }
    if (input.currencyCode !== undefined) {
      update.currency_code = input.currencyCode
    }
    if (input.openingBalance !== undefined) {
      update.opening_balance = input.openingBalance
    }
    if (input.notes !== undefined) {
      update.notes = input.notes
    }
    if (input.isActive !== undefined) {
      update.is_active = input.isActive
    }
    if (input.bankSubtype !== undefined) {
      update.bank_subtype = input.bankSubtype
    }
    if (input.creditCardLimit !== undefined) {
      update.credit_card_limit = input.creditCardLimit
    }
    if (input.dueDayOfMonth !== undefined) {
      update.due_day_of_month = input.dueDayOfMonth
    }
    if (input.investmentType !== undefined) {
      update.investment_type = input.investmentType
    }
    if (input.balanceGrams !== undefined) {
      update.balance_grams = input.balanceGrams
    }
    if (input.propertyType !== undefined) {
      update.property_type = input.propertyType
    }
    if (input.ownershipPercentage !== undefined) {
      update.ownership_percentage = input.ownershipPercentage
    }
    if (input.businessType !== undefined) {
      update.business_type = input.businessType
    }
    if (input.industry !== undefined) {
      update.industry = input.industry
    }
    if (input.location !== undefined) {
      update.location = input.location
    }
    if (input.metalType !== undefined) {
      update.metal_type = input.metalType
    }
    if (input.purity !== undefined) {
      update.purity = input.purity
    }
    if (input.purchaseDate !== undefined) {
      update.purchase_date = input.purchaseDate
    }
    if (input.costPerUnit !== undefined) {
      update.cost_per_unit = input.costPerUnit
    }

    const { data, error } = await this.client
      .from("financial_accounts")
      .update(update)
      .eq("id", id)
      .eq("user_id", userId)
      .select(accountSelect)
      .single()

    throwAccountConstraintError(error, operation)
    return mapAccountSummary(
      requireQueryData(data, error, operation),
      operation
    )
  }

  async closeAccount(id: string): Promise<AccountSummary> {
    const operation = "accounts.closeAccount"
    const { error } = await this.client.rpc("close_financial_account", {
      p_account_id: id,
    })
    requireQueryData(true, error, operation)
    return this.getAccount(id)
  }

  async reopenAccount(id: string): Promise<AccountSummary> {
    const operation = "accounts.reopenAccount"
    const { error } = await this.client.rpc("reopen_financial_account", {
      p_account_id: id,
    })
    requireQueryData(true, error, operation)
    return this.getAccount(id)
  }

  async getAccountLifecycleEligibility(
    accountIds: string[]
  ): Promise<AccountLifecycleEligibility[]> {
    const operation = "accounts.getAccountLifecycleEligibility"
    if (accountIds.length === 0) return []
    const { data, error } = await this.client.rpc(
      "get_account_lifecycle_eligibility",
      { p_account_ids: accountIds }
    )
    return requireQueryData(data, error, operation).map((row) => ({
      accountId: row.account_id,
      canClose: row.can_close,
      closeBlockReason: row.close_block_reason,
      canDelete: row.can_delete,
      deleteBlockReason: row.delete_block_reason,
      hasFinancialHistory: row.has_financial_history,
    }))
  }

  async deleteAccount(id: string): Promise<void> {
    const operation = "accounts.deleteAccount"
    const { error } = await this.client.rpc("delete_pristine_financial_account", {
      p_account_id: id,
    })
    requireQueryData(true, error, operation)
  }
}

export const accountsRepository = new AccountsRepository()

File: apps/mobile/lib/accounts/accounts_repository.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import 'account_form.dart';
import 'account_models.dart';

/// Raised for account reads/mutations; [message] is already user-facing (port of
/// the web `RepositoryError` + `createAccount`/`updateAccount` friendly-error
/// translation, docs/accounts.md §4, §5).
class AccountsException implements Exception {
  AccountsException(this.message);
  final String message;
  @override
  String toString() => message;
}

/// Every `financial_accounts` numeric column is `::text` cast so exact decimal
/// precision survives the wire (docs/accounts.md §2.4).
const _accountSelect =
    'id,user_id,account_type_code,name,currency_code,opening_balance::text,'
    'is_active,notes,bank_subtype,credit_card_limit::text,due_day_of_month,'
    'investment_type,balance_grams::text,property_type,'
    'ownership_percentage::text,business_type,industry,location,metal_type,'
    'purity,purchase_date,cost_per_unit::text,closed_reason,closed_on,'
    'created_at,updated_at';

class AccountsRepository {
  AccountsRepository([SupabaseClient? client])
    : _client = client ?? Supabase.instance.client;

  final SupabaseClient _client;
  String get _userId => _client.auth.currentUser!.id;

  // ---- reads --------------------------------------------------------------

  Future<List<Account>> getAccounts() async {
    final rows = await _client
        .from('financial_accounts')
        .select(_accountSelect)
        .eq('user_id', _userId)
        .order('created_at', ascending: false);
    return (rows as List)
        .map((r) => Account.fromRow((r as Map).cast<String, dynamic>()))
        .toList();
  }

  Future<Account> getAccount(String id) async {
    try {
      final row = await _client
          .from('financial_accounts')
          .select(_accountSelect)
          .eq('id', id)
          .eq('user_id', _userId)
          .single();
      return Account.fromRow((row).cast<String, dynamic>());
    } on PostgrestException catch (e) {
      throw AccountsException(_friendly(e));
    }
  }

  Future<List<AccountBalance>> getAccountBalances(List<String> ids) async {
    if (ids.isEmpty) return const [];
    final rows = await _client.rpc(
      'get_account_balances',
      params: {'p_account_ids': ids},
    );
    return (rows as List)
        .map((r) => AccountBalance.fromRow((r as Map).cast<String, dynamic>()))
        .toList();
  }

  Future<List<AccountValuation>> getEffectiveValuations(
    List<String> ids,
  ) async {
    if (ids.isEmpty) return const [];
    final rows = await _client.rpc(
      'get_effective_account_valuations',
      params: {'p_account_ids': ids},
    );
    // Ordered valued_on desc, created_at desc — first row per account is latest.
    return (rows as List)
        .map(
          (r) => AccountValuation.fromRow((r as Map).cast<String, dynamic>()),
        )
        .toList();
  }

  Future<List<AccountLifecycle>> getLifecycleEligibility(
    List<String> ids,
  ) async {
    if (ids.isEmpty) return const [];
    final rows = await _client.rpc(
      'get_account_lifecycle_eligibility',
      params: {'p_account_ids': ids},
    );
    return (rows as List)
        .map(
          (r) => AccountLifecycle.fromRow((r as Map).cast<String, dynamic>()),
        )
        .toList();
  }

  // ---- mutations --------------------------------------------------------

  Future<Account> createAccount(AccountFormValues v) async {
    final fields = toAccountTypeSpecificFields(v);
    final name = _accountName(v);
    try {
      if (v.type.isValued) {
        // Real Estate / Business are created atomically with their first
        // valuation via the dedicated RPC (docs/accounts.md §2.1).
        final row = await _client.rpc(
          'create_valued_account',
          params: {
            'p_account_type_code': v.type.code,
            'p_name': name,
            'p_currency_code': v.currencyCode,
            'p_property_type': fields.propertyType,
            'p_business_type': fields.businessType,
            'p_industry': fields.industry,
            'p_ownership_percentage': v.ownershipPercentage.trim(),
            'p_location': fields.location,
            'p_account_notes': _blankToNull(v.notes),
            'p_valuation_amount': v.openingBalance.trim(),
            'p_valued_on': v.valuationDate.trim(),
            // Free-text method, sent verbatim (web `p_valuation_method`).
            'p_valuation_method': _blankToNull(v.valuationMethod),
            'p_valuation_notes': _blankToNull(v.valuationNotes),
          },
        );
        final id = row is Map ? '${row['id']}' : '$row';
        return getAccount(id);
      }

      final columns = fields.toColumns(name: name, currencyCode: v.currencyCode)
        ..['user_id'] = _userId
        ..['account_type_code'] = v.type.code
        ..['notes'] = _blankToNull(v.notes);
      final row = await _client
          .from('financial_accounts')
          .insert(columns)
          .select(_accountSelect)
          .single();
      return Account.fromRow((row).cast<String, dynamic>());
    } on PostgrestException catch (e) {
      throw AccountsException(_friendly(e));
    }
  }

  Future<Account> updateAccount(String id, AccountFormValues v) async {
    final fields = toAccountTypeSpecificFields(v);
    final columns = fields.toColumns(name: _accountName(v))
      ..['notes'] = _blankToNull(v.notes);
    // currency_code / opening_balance stay editable only without history; the DB
    // triggers enforce that and _friendly surfaces the message.
    if (!v.type.isValued) columns['currency_code'] = v.currencyCode;
    try {
      final row = await _client
          .from('financial_accounts')
          .update(columns)
          .eq('id', id)
          .eq('user_id', _userId)
          .select(_accountSelect)
          .single();
      return Account.fromRow((row).cast<String, dynamic>());
    } on PostgrestException catch (e) {
      throw AccountsException(_friendly(e));
    }
  }

  Future<void> closeAccount(String id) =>
      _lifecycleRpc('close_financial_account', id);
  Future<void> reopenAccount(String id) =>
      _lifecycleRpc('reopen_financial_account', id);
  Future<void> deleteAccount(String id) =>
      _lifecycleRpc('delete_pristine_financial_account', id);

  Future<void> _lifecycleRpc(String fn, String id) async {
    try {
      await _client.rpc(fn, params: {'p_account_id': id});
    } on PostgrestException catch (e) {
      throw AccountsException(_friendly(e));
    }
  }

  /// Gold/silver accounts are always auto-named "Gold" / "Silver" (§4, §9.1).
  static String _accountName(AccountFormValues v) {
    if (v.type == AccountType.gold) {
      return v.metalType == 'silver' ? 'Silver' : 'Gold';
    }
    return v.name.trim();
  }

  static Object? _blankToNull(String? s) =>
      (s == null || s.trim().isEmpty) ? null : s.trim();

  static String _friendly(PostgrestException e) {
    final code = e.code ?? '';
    final message = e.message;
    if (code == 'PGRST116' ||
        code == 'P0002' ||
        message.contains('account not found')) {
      return 'That account is no longer available.';
    }
    if (code == '23505') {
      if (message.contains('user_currency_metal_type')) {
        return 'You already have this type of Gold/Silver account in this '
            'currency. Go to that account and add a purchase instead of '
            'creating a new one.';
      }
      return 'An account with this name already exists.';
    }
    if (code == '23514') {
      if (message.contains('account_close_blocked')) {
        return _lifecycleReason(message, 'close');
      }
      if (message.contains('account_delete_blocked')) {
        return _lifecycleReason(message, 'delete');
      }
      if (message.contains('account_reopen_blocked')) {
        return 'A sold account can’t be reopened.';
      }
      if (message.contains('currency cannot be changed')) {
        return 'This account already has financial history — its currency '
            'can’t be changed.';
      }
      if (message.contains('opening balance')) {
        return 'This account already has financial history — its opening '
            'balance can’t be changed.';
      }
      if (message.contains('ownership')) {
        return 'This account already has history — ownership is managed by the '
            'disposal timeline.';
      }
      return 'That change isn’t allowed for this account.';
    }
    if (code == '42501') return 'You don’t have permission to do that.';
    return 'Something went wrong. Please try again.';
  }

  static String _lifecycleReason(String message, String verb) {
    final reason = message.split(':').last.trim();
    final human = switch (reason) {
      'nonzero_balance' || 'nonzero_value' => 'the balance must be zero first',
      'nonzero_amount_due' => 'the amount due must be zero first',
      'remaining_metal_quantity' => 'sell or move the remaining metal first',
      'has_positive_holdings' ||
      'brokerage_has_value' => 'clear the holdings and available cash first',
      'ownership_remaining' ||
      'real_estate_business' => 'use the full-sale flow instead',
      'has_financial_history' ||
      'not_pristine' => 'it already has financial history',
      _ => 'it isn’t eligible',
    };
    return 'This account can’t be ${verb}d — $human.';
  }
}
